### PR TITLE
[FlashScript] OOP version, initial commit.

### PR DESCRIPTION
**Pull request by andyblac**

### CODE REVIEW

**Comment by timoschilling** [comment]
I assume that this is the PR for this build? https://github.com/mongoose-os-apps/shelly-homekit/issues/314#issuecomment-738239212

If yes I would like to give some feedback:
1. The devise id of stock devises should not contain `.local` as it is by HomeKit devises
2. Maybe you could prefix the host with `http://` that will allow to open the web ui via click from inside the console. In my setup with an unconfigured Safari and an unconfigured iTerm2 terminal it will open all links as `https://` as default. This change will be help to enforce the right `http://` protocol handler.

But at all great work!

**Comment by andyblac** [comment]
didn't know about url links, i don't use iTerm2, and links do not work in default Terminal app.

**Comment by andyblac** [comment]
just tried iterm2 links do not seem to work here.

**Comment by andyblac** [comment]
ahh you have to hold CMD down and click on url

File: tools/flash-shelly.py [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+  sorted_device_list = sorted(device_list, key=lambda k: k['host'])
+  logger.trace(f"device_test: {sorted_device_list}")
+
+  for device in sorted_device_list:
+    if mode == 'keep':
+      flashmode = device['fw_type']
+    else:
+      flashmode = mode
+
+    if device['fw_type'] == 'homekit':
```
Comment (rojer) [inline tools/flash-shelly.py:568]: nice, i like this part, much better than it used to be!

**Review by rojer — COMMENTED** [review]
overall looks fine. i haven't done detailed review, but overall seems better than before.

one nitpick: there are three common styles for naming functions: `lower_case_underscore_style()`, `camelCaseStyle()` and `CapitalizedcamelCase()`. amazingly, you managed to use them all :) 
be consistent: pick one and stick to it. fwiw, python style guide recommends `lower_case_underscore_style()`.

File: tools/flash-shelly.py (hunk 1 above)
Comment (andyblac) [reply tools/flash-shelly.py:568]: ok done changed to `lower_case_underscore_style()`

File: tools/flash-shelly.py [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+    if not self.version:
+      if self.stock_model  == 'SHRGBW2':
+        self.dlurl = stock_model_info['url'].replace('.zip',f'-{self.colour_mode}.zip')
+      else:
+        self.dlurl = stock_model_info['url']
+    else:
+      self.dlurl = f'http://archive.shelly-faq.de/version/v{self.version}/{self.stock_model}.zip'
+
+class HomeKitDevice(Device):
+  def __init__(self, host, wifi_ip, fw_type, device_url, info, variant, version):
```
Comment (rojer) [inline tools/flash-shelly.py:259]: no need to define `__init__` if you're not overriding anything

File: tools/flash-shelly.py [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+    myfile = requests.get(self.dlurl)
+    logger.info("Now Flashing...")
+    files = {'file': ('shelly-flash.zip', myfile.content)}
+    logger.debug(f"requests.post(f'http://{self.wifi_ip}/update' , files=files")
+    response = requests.post(f'http://{self.wifi_ip}/update' , files=files)
+    logger.debug(response.text)
+
+
+class StockDevice(Device):
+  def __init__(self, host, wifi_ip, fw_type, device_url, info, variant, version):
```
Comment (rojer) [inline tools/flash-shelly.py:300]: same here

File: tools/flash-shelly.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+  def add_service(self, zeroconf, type, device):
+    device = device.replace('._http._tcp.local.', '')
+    deviceinfo = Device(device)
+    deviceinfo.get_device_info()
+    if deviceinfo.fw_type is not None:
+      dict = {'host': deviceinfo.host, 'wifi_ip': deviceinfo.wifi_ip, 'fw_type': deviceinfo.fw_type, 'device_url': deviceinfo.device_url, 'info' : deviceinfo.info}
+      self.device_list.append(dict)
+
+class Device():
```
Comment (rojer) [inline tools/flash-shelly.py:124]: remove parens, don't need them

File: tools/flash-shelly.py (hunk 3 above)
Comment (andyblac) [reply tools/flash-shelly.py]: done

File: tools/flash-shelly.py (hunk 2 above)
Comment (andyblac) [reply tools/flash-shelly.py]: done

File: tools/flash-shelly.py (hunk 4 above)
Comment (andyblac) [reply tools/flash-shelly.py]: done

**Pull request merged**